x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java

```diff
@@ -298,6 +293,9 @@ public DatafeedConfig apply(DatafeedConfig datafeedConfig, Map<String, String> h
 
         DatafeedConfig.Builder builder = new DatafeedConfig.Builder(datafeedConfig);
         if (jobId != null) {
+            if (datafeedConfig.getJobId() != null && datafeedConfig.getJobId().equals(jobId) == false) {
+                throw ValidationException.withError(ERROR_MESSAGE_ON_JOB_ID_UPDATE);
```

Comment:
A possible problem with using `ValidationException` here is that this method is being called on the master node but the original request is likely to have been sent to some other node.  This exception will get wrapped in a `RemoteTransportException` when it is sent back to the coordinating node.  There is logic to determine the "root cause" of an exception, and the rule is that it's the most deeply nested exception that inherits from `ElasticsearchException`.  Since `RemoteTransportException` inherits from `ElasticsearchException` and `ValidationException` doesn't, this means the `RemoteTransportException` will be reported as the root cause.

Exceptions that are root causes but do not inherit from `ElasticsearchException` (such as `IllegalArgumentException` or `ValidationException`) should only be used in places where they'll be thrown on the coordinating node.  Any exceptions that are root causes and will be thrown on a different node must extend `ElasticsearchException`.

TL;DR it's probably better to use a `ElasticsearchStatusException` here, which you can do by calling `org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper.badRequestException`.

Reply:
Done.

Thanks for detailed explanation of exception handling!